### PR TITLE
CBMC Proofs for MQTT_Init, MQTT_GetPacketId, and MQTT_GetIncomingPacketTypeAndLength

### PR DESCRIPTION
**Pull request by sarenameas**
*PLEASE READ ALL Description of changes:*

Proofs for:
- MQTT_Init
- MQTT_GetPacketId
- MQTT_GetIncomingPacketTypeAndLength

The symbolically linked files in PR (Makefile-project-defines, Makefile-project-targets, Makefile-project-testing, Makefile-template-defines, README.mds, etc) make up the CBMC infrastructure. They are linked from the new submodule pulled in. This helps the CBMC team make infrastructure changes without creating PRs to this repo.

**This PR depends on https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1014 for the CBMC proofs to pass.** 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/mqtt/cbmc/proofs/Makefile-project-targets [hunk 1]
```diff
@@ -0,0 +1 @@
+../../../../../tools/aws-templates-for-cbmc-proofs/template-for-repository/proofs/Makefile-project-targets
```
Comment (sarenameas) [inline libraries/standard/mqtt/cbmc/proofs/Makefile-project-targets:1]: Per offline conversation we shall instruct users to run setup.py in the template submodule.

File: docs/cbmc.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+1. Change the current working directory to the library's cbmc folder:  
+   `cd <SRC_ROOT>/libraries/standard/<LIBRARY_NAME>/cbmc`
+
+1. Run the setup.py script in aws-templates-for-cbmc-proofs with Python 3.6+:  
+   `Python3 ../../../../aws-templates-for-cbmc-proofs/scripts/setup.py`  
+   When prompted for the source root set either the relative path or the absolute path:  
+   `What is the path to the source root: ../../../..`
+
+Setting up a new proof and other information about the aws-templates-for-cbmc-proofs can be found at https://github.com/awslabs/aws-templates-for-cbmc-proofs.
```
Comment (sarenameas) [inline docs/cbmc.md:15]: How to download and run CBMC is to be verified first for the public README.md (we have internal instructions on our compant wiki) and added later.

File: libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# Preprocessor include paths -I...
+INCLUDES += \
+        -I$(SRCDIR)/libraries/standard/mqtt/cbmc/include \
+        -I$(SRCDIR)/libraries/standard/mqtt/include \
+        -I$(SRCDIR)/libraries/standard/utilities/include \
+        -I$(SRCDIR)/libraries/standard/mqtt/src \
+        -I$(SRCDIR)/demos/logging-stack \
+        -I$(SRCDIR)/demos/mqtt/mqtt_demo_basic_tls \
+        -I$(SRCDIR)/demos/mqtt/mqtt_demo_lightweight \
+        -I$(SRCDIR)/demos/mqtt/mqtt_demo_plaintext
```
Comment (yourslab) [inline libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines:30]: I think we should be able to include the `mqtt_demo_plaintext` demo only or have a separate config header file for these proofs.

File: libraries/standard/mqtt/cbmc/proofs/Makefile-template-defines [hunk 4]
```diff
@@ -0,0 +1 @@
+SRCDIR ?= $(abspath $(PROOF_ROOT)/../../../../..)
```
Comment (yourslab) [inline libraries/standard/mqtt/cbmc/proofs/Makefile-template-defines:1]: I think this file is not required and is generated by `setup.py`

File: docs/cbmc.md (hunk 2 above)
Comment (yourslab) [reply docs/cbmc.md:15]: Nice!

File: libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines (hunk 3 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines:30]: yes

File: libraries/standard/mqtt/cbmc/proofs/Makefile-template-defines (hunk 4 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/Makefile-template-defines:1]: i didnt delete it because it wasn't linked

File: docs/cbmc.md (hunk 2 above)
Comment (markrtuttle) [reply docs/cbmc.md:15]: None of these instructions should be required now that the templates repository is submoduled and will be grabbed by the normal git submodule update, and now that the symbolic links are part of the current repository.  The only thing that might be necessary is how to install cbmc and cbmc-viewer, and how to run setup-proofs.py to start a new proof.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Insert copyright notice
+ */
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:3]: I think the copyright stuff is required on all these files before putting into open source.

```
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
```

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * @file MQTT_GetIncomingPacketTypeAndLength_harness.c
+ * @brief Implements the proof harness for MQTT_GetIncomingPacketTypeAndLength function.
+ */
+#include "mqtt.h"
+
+extern int32_t NetworkInterfaceReceiveStub( NetworkContext_t context,
+                                            void * pBuffer,
+                                            size_t bytesToRecv );
+
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:14]: I'd create a NetworkInterfaceReceiveStub.h in the cbmc/includes folder.  I'll bet you are going to use this stub a lot of places.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/Makefile [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+INCLUDES +=
+
+REMOVE_FUNCTION_BODY +=
+UNWINDSET += getRemainingLength.0:5
+
+PROOF_SOURCES += $(PROOFDIR)/$(HARNESS_FILE).c
+PROOF_SOURCES += $(PROOFDIR)/../../stubs/NetworkInterfaceReceiveStub.c
+PROJECT_SOURCES += $(SRCDIR)/libraries/standard/mqtt/src/mqtt_lightweight.c
+
+include ../Makefile.common
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/Makefile:14]: All files should end with a newline.

And copyright

```
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
```

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+void harness()
+{
+    NetworkContext_t networkContext;
+
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:18]: I'm surprised there is nothing more to allocation than just a pointer.  No backing struct?

File: .gitignore [hunk 9]
```diff
@@ -5,10 +5,20 @@ doc/tag/*
...
 *.gcda
 *.gcno
 *.gcov
+
+# Ignore CBMC proof artifacts
+**/cbmc/proofs/**/gotos
+**/cbmc/proofs/**/html
+**/cbmc/proofs/**/logs
+**/cbmc/proofs/**/TAGS-*
+**/cbmc/**/README.md
```
Comment (markrtuttle) [inline .gitignore:21]: I certainly would not ignore the README files.

File: .gitignore [hunk 10]
```diff
@@ -5,10 +5,20 @@ doc/tag/*
...
+**/cbmc/proofs/Makefile.common
+**/cbmc/proofs/Makefile-project-targets
+**/cbmc/proofs/Makefile-project-testing
```
Comment (markrtuttle) [inline .gitignore:24]: I don't think you can ignore these files.  If they are symbolic links to the templates repository, then we need the symbolic links.  If they are actual files replacing the symbolic links, then we need the actual files.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    __CPROVER_assume( networkContext != NULL );
+
+    MQTTPacketInfo_t incomingPacketInfo;
+
+    MQTT_GetIncomingPacketTypeAndLength( NetworkInterfaceReceiveStub,
+                                         networkContext,
+                                         &incomingPacketInfo );
+}
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:26]: This is beyond memory safety, but is there anything about the data written to incomingPacketInfo that we could test?  Like lengths or pointers nonnull or something?

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/cbmc-batch.yaml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+# This file marks this directory as containing a CBMC proof.
+# The contents of this file will be generated by continuous integration.
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/cbmc-batch.yaml:2]: Any idea why these files do not end with a newline?  I'm pretty sure they exist in the templates.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetPacketId/MQTT_GetPacketId_harness.c [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * @file MQTT_GetPacketId_harness.c
+ * @brief Implements the proof harness for MQTT_GetPacketId function.
+ */
+
+#include "mqtt.h"
+
+void harness()
+{
+    MQTTContext_t context;
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetPacketId/MQTT_GetPacketId_harness.c:14]: Let's investigate the malloc flag for cbmc and consider mallocing a possibly null pointer to a MQTTContext_t.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_Init/MQTT_Init_harness.c [hunk 14]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Insert copyright notice
+ */
+
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_Init/MQTT_Init_harness.c:4]: copyright

File: libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+# binaries and cbmc to do the property and coverage checking.
+#
+# Use this file to override most default definitions of variables in
+# Makefile.common.
+################################################################
+
+# Flags to pass to goto-cc for compilation (typically those passed to gcc -c)
+COMPILE_FLAGS += \
+        -fPIC \
+        -std=gnu90
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines:16]: Please consider using the += style.  I saw some evidence with HTTP that the line continuation style does not interact with the new Makefile.common macros that my team has developed.

```
COMPILE_FLAGS += -fPIC 
COMPILE_FLAGS += -std=gnu90

INCLUDES += -I$(SRCDIR)/libraries/standard/mqtt/cbmc/include 
INCLUDES += -I$(SRCDIR)/libraries/standard/mqtt/include 
INCLUDES += -I$(SRCDIR)/libraries/standard/utilities/include 
INCLUDES += -I$(SRCDIR)/libraries/standard/mqtt/src 
INCLUDES += -I$(SRCDIR)/demos/logging-stack 
INCLUDES += -I$(SRCDIR)/demos/mqtt/mqtt_demo_plaintext

DEFINES += -Dmqtt_EXPORTS
```

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c [hunk 16]
```diff
@@ -0,0 +1,19 @@
+#include "mqtt.h"
+
+int32_t NetworkInterfaceReceiveStub( NetworkContext_t context,
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:5]: Another place the header file might be useful.

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    __CPROVER_assert( context != NULL,
+                      "IotNetworkInterfaceReceive pConnection is not NULL." );
+
+    __CPROVER_assert( pBuffer != NULL,
+                      "IotNetworkInterfaceReceive pBuffer is not NULL." );
+    __CPROVER_havoc_object( pBuffer );
+
+    /* Choose the number of bytes in pBuffer considered received. */
+    int32_t bytesOrError;
+    __CPROVER_assume( bytesOrError <= ( int32_t ) bytesToRecv );
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:16]: Is there no need to assume that bytesOrError is >= 0?

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
+{
+    __CPROVER_assert( context != NULL,
+                      "IotNetworkInterfaceReceive pConnection is not NULL." );
+
+    __CPROVER_assert( pBuffer != NULL,
+                      "IotNetworkInterfaceReceive pBuffer is not NULL." );
+    __CPROVER_havoc_object( pBuffer );
+
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:13]: Should we be asserting that pBuffer points to a buffer with at least bytesToRecv characters in it?

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+int32_t NetworkInterfaceReceiveStub( NetworkContext_t context,
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
+{
+    __CPROVER_assert( context != NULL,
+                      "IotNetworkInterfaceReceive pConnection is not NULL." );
+
+    __CPROVER_assert( pBuffer != NULL,
+                      "IotNetworkInterfaceReceive pBuffer is not NULL." );
+    __CPROVER_havoc_object( pBuffer );
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:12]: Nicely done.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/cbmc-batch.yaml (hunk 12 above)
Comment (yourslab) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/cbmc-batch.yaml:2]: They do not seem to end with a newline by default for me either.

File: docs/cbmc.md (hunk 2 above)
Comment (sarenameas) [reply docs/cbmc.md:15]: Ok we shall delete these in anticipation of the sub-module going public.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 5 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:3]: I did not add it yet as the license for our code is not finalized yet. I can add this in temporarily.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 6 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:14]: It appears I will use it in two more placed, but I will move it to a header nonetheless since it is more than once.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/Makefile (hunk 7 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/Makefile:14]: True, but this is also auto-generated, so the script needs to make it end in a new line?

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 8 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:18]: No because this is an application defined structure. I had comments there at first, I will add them back.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 11 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:26]: No there is not, I had comments about it, and i will add them back.

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c (hunk 17 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:16]: No we want to take the error path in the code. Well I just thought about it and the MQTT code should also handle greater than bytesToRecv, so i will remove this.

File: .gitignore (hunk 10 above)
Comment (sarenameas) [reply .gitignore:24]: Per offline conversation we shall check in the symbolic links.

File: .gitignore (hunk 9 above)
Comment (sarenameas) [reply .gitignore:21]: They were symbolically linked.

File: libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines (hunk 15 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/Makefile-project-defines:16]: OK

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c (hunk 19 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:34]: ty

File: libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c (hunk 18 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/stubs/NetworkInterfaceReceiveStub.c:35]: How do we do that?

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /* NetworkContext_t is an application defined network interface context. It
+     * is passed through to the readFunc parameter of 
+     * MQTT_GetIncomingPacketTypeAndLength(). */
+    NetworkContext_t networkContext;
+
+    __CPROVER_assume( networkContext != NULL );
+
+    /* MQTT_GetIncomingPacketTypeAndLength() will set only the remainingLength
+     * field in in the input MQTTPacketInfo_t structure. */
```
Comment (yourslab) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:40]: Double in

File: libraries/standard/mqtt/cbmc/sources/mqtt_cbmc_state.c [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include "mqtt_cbmc_state.h"
+
+void * mallocCanFail( size_t size )
+{
+    __CPROVER_assert( size < CBMC_MAX_OBJECT_SIZE, "mallocCanFail size is too big" );
+    return nondet_bool() ? NULL : malloc( size );
+}
```
Comment (yourslab) [inline libraries/standard/mqtt/cbmc/sources/mqtt_cbmc_state.c:30]: Pls add newline

File: libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * Proof model for malloc that can fail and return NULL.
+ * 
+ * @param[in] size The size in bytes of memory to allocate.
+ * @return NULL or requested memory.
+ */
+void * mallocCanFail( size_t size );
```
Comment (yourslab) [inline libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h:28]: Newline too

File: libraries/standard/mqtt/cbmc/sources/mqtt_cbmc_state.c (hunk 21 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/sources/mqtt_cbmc_state.c:30]: Thanks I was tired yesterday and didnt want to look at the diffs until today haha

File: libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h (hunk 22 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h:28]: Havent run uncrustify yet because got tired

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 20 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:40]: Thanks

File: libraries/standard/mqtt/src/mqtt.c [hunk 23]
```diff
@@ -1468,9 +1468,12 @@ uint16_t MQTT_GetPacketId( MQTTContext_t * pContext )
     if( pContext != NULL )
     {
         packetId = pContext->nextPacketId;
-        pContext->nextPacketId++;
 
-        if( pContext->nextPacketId == 0U )
+        if( pContext->nextPacketId == UINT16_MAX )
```
Comment (sarenameas) [inline libraries/standard/mqtt/src/mqtt.c:1472]: These cahnges will be moved to a separate PR and documented.

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 24]
```diff
@@ -1292,6 +1307,51 @@ static void serializeConnectPacket( const MQTTConnectInfo_t * pConnectInfo,
 
 /*-----------------------------------------------------------*/
 
+MQTTStatus_t getIncomingPacketTypeAndLength( MQTTTransportRecvFunc_t readFunc,
```
Comment (sarenameas) [inline libraries/standard/mqtt/src/mqtt_lightweight.c:1310]: Moved to separate function for complexity.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file MQTT_GetIncomingPacketTypeAndLength_harness.c
+ * @brief Implements the proof harness for MQTT_GetIncomingPacketTypeAndLength function.
+ */
+#include "mqtt.h"
+#include "NetworkInterfaceReceiveStub.h"
```
Comment (yourslab) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:27]: I think this should be underscore'd for consistency

File: .gitignore [hunk 26]
```diff
@@ -5,10 +5,16 @@ doc/tag/*
...
 # Ignore code coverage artifacts
 *.gcda
 *.gcno
 *.gcov
+
+# Ignore CBMC proof artifacts
+**/cbmc/proofs/**/gotos
+**/cbmc/proofs/**/html
+**/cbmc/proofs/**/logs
+**/cbmc/proofs/**/TAGS-*
```
Comment (markrtuttle) [inline .gitignore:20]: Could add `**/cbmc/proofs/**/report` which is the equivalent of html for the new viewer.

File: libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @brief Proof model for malloc that can fail and return NULL.
+ *
+ * @param[in] size The size in bytes of memory to allocate.
+ * @return NULL or requested memory.
+ */
+void * mallocCanFail( size_t size );
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h:28]: Did you move from malloc_can_fail to mallocCanFail?

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+    /* NetworkContext_t is an application defined network interface context. It
+     * is passed through to the readFunc parameter of
+     * MQTT_GetIncomingPacketTypeAndLength(). */
+    NetworkContext_t networkContext;
+
+    __CPROVER_assume( networkContext != NULL );
+
+    /* MQTT_GetIncomingPacketTypeAndLength() will set only the remainingLength
+     * field in the input MQTTPacketInfo_t structure. */
+    MQTTPacketInfo_t * pIncomingPacketInfo = mallocCanFail( sizeof( MQTTPacketInfo_t ) );
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:41]: At some point in the future you will probably want an allocator for a packet info.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+#include "mqtt_cbmc_state.h"
+
+void harness()
+{
+    /* NetworkContext_t is an application defined network interface context. It
+     * is passed through to the readFunc parameter of
+     * MQTT_GetIncomingPacketTypeAndLength(). */
+    NetworkContext_t networkContext;
+
+    __CPROVER_assume( networkContext != NULL );
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:37]: If this is an application-defined pointer that is never dereferenced by your code, you probably don't need to assume nonnull.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetPacketId/MQTT_GetPacketId_harness.c [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+#include "mqtt.h"
+#include "mqtt_cbmc_state.h"
+
+void harness()
+{
+    /* The MQTTContext_t is allocated such that we can test a NULL input.
+     * MQTT_GetPacketId() touches only the nextPacketId field in MQTTContext_t.
+     * This nextPacketId is left unbounded to verify the function under harness.
+     */
+    MQTTContext_t * pContext = mallocCanFail( sizeof( MQTTContext_t ) );
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_GetPacketId/MQTT_GetPacketId_harness.c:36]: At some point in the future you may want to add a context allocator and is_valid predicate.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_Init/MQTT_Init_harness.c [hunk 31]
```diff
@@ -0,0 +1,40 @@
...
+    MQTTContext_t context;
+    MQTTTransportInterface_t transportInterface;
+    MQTTApplicationCallbacks_t callbacks;
+    MQTTFixedBuffer_t networkBuffer;
+
+    MQTT_Init( nondet_bool() ? NULL : &context,
+               nondet_bool() ? NULL : &transportInterface,
+               nondet_bool() ? NULL : &callbacks,
+               nondet_bool() ? NULL : &networkBuffer );
+}
```
Comment (markrtuttle) [inline libraries/standard/mqtt/cbmc/proofs/MQTT_Init/MQTT_Init_harness.c:40]: This is an example of why you might want allocators for these things.  An allocator would call mallocCanFail which could return either null or a valid pointer.  As you write more proofs, you may want to revisit this.

**Comment by markrtuttle** [comment]
I didn't understand the comment about git turning symlinks into actual files.  I haven't observed that.  Strangely, however, I am not unable to checkout your branch on a Mac because MacOS complains about "symlink too long"!  Unbelievable.

File: libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h (hunk 27 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/include/mqtt_cbmc_state.h:28]: Yes our coding style lol

**Comment by sarenameas** [comment]
The most recent comments will appear in the next PR.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 28 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:41]: Yes that will def come for the functions that use packetInfo more intensely

File: .gitignore (hunk 26 above)
Comment (sarenameas) [reply .gitignore:20]: Added in mqtt/proofs branch. Will be in the next PR.

File: libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c (hunk 29 above)
Comment (sarenameas) [reply libraries/standard/mqtt/cbmc/proofs/MQTT_GetIncomingPacketTypeAndLength/MQTT_GetIncomingPacketTypeAndLength_harness.c:37]: Thanks
